### DISCUSSION TITLE
Batocera Lite/Koriki Alpha for RG35XX Plus - 20240122

### DISCUSSION THREAD

**Discussion by acmeplus** [rg35xx-cfw/rg35xx-cfw.github.io#77] [Announcements]
### This is an early alpha image of batocera Lite/koriki for the RG35XX Plus, things are starting to work, but there many features missing at this point.

Important: there are still several performance issues with this release, while there are cores/emulators that are running fine, some others are not there yet. In general the build needs to be updated with optimizations, but there's still work to do.

### New install:
* Flash image to a new sdcard
* Insert sdcard into the RG35XX Plus
* Press power a few seconds to boot
* NOTE: The first time the system boots it will take a bit longer while the SDCARD is getting expanded and the initial SHARE folder content gets populated

### Update from a previous version (2023/12/17 or higher):
* Extract the boot.tar.xz file
* Mount the SDCARD on your computer
* Delete the file ``boot/batocera`` from the first partition of the SDCARD (BATOCERA)
* Copy its contents to the third partition of the SDCARD (boot-resource or BOOT)
* Insert the SDCARD in your RG35XX plus and boot

## Cores/Emulators Status:
* Working: amstrad, mame, atari2600, atari5200, atari7800, wonderswan color, amiga(uae4arm), c64 (libretro x64), pico8, scummvm (core), vectrex, dosbox, pcengine, pcenginecd, g&w, nes, gb, snes, gbc, gba, openbor, sg-1000, mastersystem, megadrive, gamegear, spectrum, neogeo, neocd, ngpc, psx, psp, odyssey2, drastic, n64 (.zip with parallel libretro core, z64/n64 with mupen64plus)
* Broken: scummvm standalone (unless using an external mouse), colecovision (missing core), Intellivision (probably a bios issue), msx2 (core missing), megacd, x86k (bios issue), flycast standalone

### 2024/01/22:
* Fixed shutdown, reboot. Thanks @xonglebongle (adixal) for the hint
* Switched back to GCC compilation, all cores recompiled
* Fixed flycast standalone freezing before game starts
* Fixed Drastic (NDS) not starting
* Added workaround for Drastic control mapping. Add the file ``es_systems_nds.cfg`` to ``[SHARE]/system/configs/emulationstation/``. To use the stylus keep pressed ``R2`` to use DPAD for stylus and A to "click". 
* Confirmed ADB is working
* Fixed Retroarchievements from EmulationStation (you need to have retroarchievements.org account)
* Fixed scrapers 

### 2024/01/19:
* Fixed SHARE partition automatic expansion for single SDCard configuration
* Change SHARE format to expat
* Added additional card filesystem consistency checks for both internal and external SHARE partitions 
* Fixed SSH credentials, use root/linux as default
* Fixed flycast_libretro (dreamcast, Naomi, atomiswave) and made core selection available from emulationstation. If you applied the 2024/01/16 es_system_dreamcast.cfg workaround, remove it. 
* Flycast standalone is still broken, once you are in dreamcast, Naomi, or atomiswave section, press ``SELECT`` to open the ``VIEW OPTIONS`` menu, then go to ``ADVANCED SYSTEM OPTIONS``, and in ``EMULATOR`` choose ``LIBRETRO: FLYCAST``
* Initial bluetooth work (still not usable via frontend)
* Added some GLES3 cores/emulators: mupen64plus-next and Yabasanshiro (Saturn).
 
### 2024/01/16:
* Fixed audio click noise (not perfect but now very subtle)
* Fixed audio volume controls (volume can be controlled with the keys and is saved during reboots)
* Fixed audio not working on some emulators/cores
* Added brightness controls (level is saved across reboots)
* Added samba support. You can mount the SHARE partition as a network drive

### 2023/12/20:
* Fixed Overclock settings that were not correctly setup. In EmulationStation, change default setting in emulationstation settings
* mupen64 now working
* Added additional cores (gearcoleco, gem, race, vm, bk, gearbox, meteor, quicknes, cannonball, mame2003-xtreme, crocosd, supafaust, etc.). Note that not all cores are mapped in EmulationStation yet, in some cases you will need to open retroarch, and select the core manually to test.

## Things that don't work:
* Bluetooth
* ~~Brightness control~~
* ~~Audio has a static click sound when starting/stopping an emulator~~
* Some cores/Standalone emulators:
  * ~~Flycast~~
  * ~~Drastic (requires a minor update)~~
  * Amiga performance is not there yet
* ~~Controls in ES are not mapped initially, just press a button and map them using ES mapping tool~~
* HDMI (not tested)
* Sleep/Supend
* ~~SDCard auto-expansion does not work at this time. You need to use two SDCards at this point~~
* Boot time needs to be improved
* ~~Error 256 when launching roms. If you get that error mount the sdcard on your computer and create a folder system/logs on the last partition of the SDcard (SHARE).~~
* Batocera bezels (not tested)
* Menu language change (graphical themes get updated, but the menu remains in English)
* ~~Drastic works but stylus mapping is not correct~~
* ~~Retroachivements via ES. You can set them up via configuration though,[ check this post for instructions](https://github.com/rg35xx-cfw/rg35xx-cfw.github.io/discussions/69#discussioncomment-8164706)~~

## Things that work:
* Wifi
* Controls
* Many cores/emulators
* EmulationStation
* Simplemenu
* Brigthness control (Menu + Volume keys)
* ADB

## Additional features:
* Switching default launcher (ES/Batocera, Simplemenu/Koriki, or Retroarch):
  * Create a new file in ``[SHARE]/system/customlauncher`` with the text ``simplemenu`` in it.  
  * The next time the system starts will launch simplemenu instead of EmulationStation
  * Possible values for the custom launcher file are: ``simplemenu``, ``emulationstation``, and ``retroarch``
* One or two SDCARDs?
  * Since SDCARD SHARE partition expansion does not work at this time, the best solution is to use a second sdcard for content.
  * If you want to use just one SDCARD, you will need to expand the last partition of the SDCARD (partition 4) to use the maximum space available, not that modifying the partition table of the batocera card may break the internal bootloader, so if the sdcard does not boot after expanding the card, refresh the image and use a second card for content.
* General performance:
  * Expect some cores/emulators to perform better than the original RG35XX, but don't be surprised if some of them perform worse at this point. Cores and emulators have not been optimized yet

Use [this discussion page](https://github.com/rg35xx-cfw/rg35xx-cfw.github.io/discussions/69) for any comments, questions, or general reporting about this release.

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/A0A1J951S)

<hr /><em>This discussion was created from the release <a href='https://github.com/rg35xx-cfw/rg35xx-cfw.github.io/releases/tag/rg35xx_plus_batocera_lite_koriki_alpha_20240122'>Batocera Lite/Koriki Alpha for RG35XX Plus - 20240122</a>.</em>

**Comment by alojzjakob** [comment]
Thank you for the hard work! It looks beautiful!
Can we use the actual menu button on the device to do "menu" things, this is not RGB30 :)
We also need a battery indicator

**Reply by KingLJN** [reply]
It seems that's just how Batocera UI was designed, with the "MENU" key being a hotkey modifier instead of its own menu button. Maybe we can change that in RetroArch settings?
https://wiki.batocera.org/basic_commands

**Comment by dopazz** [comment]
I wrote the image to a 128GB SD card, popped it in my RG35XX H, and fired it up.  It auto-expanded the partition as expected and is working great.  Been using sftp to push ROMs via the wireless network, it seems to work a bit faster than the Samba Windows client.  I read in a different comment that you are going to try to add RG35XX H analog support once you receive your own; can the analog support be added through config file changes or does it require rebuilding binaries?

Thank you!

**Reply by dopazz** [reply]
My comment above was regarding the batocera-h5-rg35xx-plus--20240123 release.  I have flashed batocera-h5-rg35xx-plus--20240124 to a new card and my experience mirrors rendar82's comments below: out of the box the controls were mapped completely wrong.  I edited es_last_input.cfg and es_input.cfg as they described and I was able to move around ES using the analog sticks.  I'm copying ROMs to the new card now and will test further later.

**Comment by rendar82** [comment]
I've been running the last few versions on my RG35XX-H, they have always been able to boot but analog stick/L3 & R3 buttons were unmappable even when the axis' & buttons were manually added to the es_inpug.cfg file.

With the new image published today specifically for the RG35XX-H, I was finally able to get everything working perfectly including all analog stick controls.

Upon first boot after flashing the image, all of the controls were severely messed up. The a/b/x/y face buttons were acting as d-pad, d-pad was unresponsive, right-stick x-axis was acting as the left stick, and several keys were unmapped including the start button. In order to fix this, I had to fully delete the /userdata/system/configs/drastic/config/es_input.cfg file and reboot. After restarting, the controls were still pretty messed up but the right trigger at least acted as the start button and I was able to navigate to the control mapping tool and once running through the tool, all inputs were able to be properly mapped including the left/right analog sticks and L3/R3 buttons.

After proper remapping, here were the contents of the es_last_input.cfg & es_input.cfg file:

<inputConfig type="joystick" deviceName="Deeplay-keys" deviceGUID="19000000010000000100000000010000">
		<input name="a" type="button" id="3" value="1" code="304" />
		<input name="b" type="button" id="4" value="1" code="305" />
		<input name="down" type="hat" id="0" value="4" />
		<input name="hotkey" type="button" id="11" value="1" code="312" />
		<input name="joystick1left" type="axis" id="0" value="-1" code="2" />
		<input name="joystick1up" type="axis" id="1" value="-1" code="3" />
		<input name="joystick2left" type="axis" id="2" value="-1" code="4" />
		<input name="joystick2up" type="axis" id="3" value="-1" code="5" />
		<input name="l2" type="button" id="13" value="1" code="314" />
		<input name="l3" type="button" id="12" value="1" code="313" />
		<input name="left" type="hat" id="0" value="8" />
		<input name="pagedown" type="button" id="8" value="1" code="309" />
		<input name="pageup" type="button" id="7" value="1" code="308" />
		<input name="r2" type="button" id="14" value="1" code="315" />
		<input name="r3" type="button" id="15" value="1" code="316" />
		<input name="right" type="hat" id="0" value="2" />
		<input name="select" type="button" id="9" value="1" code="310" />
		<input name="start" type="button" id="10" value="1" code="311" />
		<input name="up" type="hat" id="0" value="1" />
		<input name="x" type="button" id="6" value="1" code="307" />
		<input name="y" type="button" id="5" value="1" code="306" />
	</inputConfig>

**Reply by rendar82** [reply]
`
<inputConfig type="joystick" deviceName="Deeplay-keys" deviceGUID="19000000010000000100000000010000">`
`
		<input name="a" type="button" id="3" value="1" code="304" />
		<input name="b" type="button" id="4" value="1" code="305" />
		<input name="down" type="hat" id="0" value="4" />
		<input name="hotkey" type="button" id="11" value="1" code="312" />
		<input name="joystick1left" type="axis" id="0" value="-1" code="2" />
		<input name="joystick1up" type="axis" id="1" value="-1" code="3" />
		<input name="joystick2left" type="axis" id="2" value="-1" code="4" />
		<input name="joystick2up" type="axis" id="3" value="-1" code="5" />
		<input name="l2" type="button" id="13" value="1" code="314" />
		<input name="l3" type="button" id="12" value="1" code="313" />
		<input name="left" type="hat" id="0" value="8" />
		<input name="pagedown" type="button" id="8" value="1" code="309" />
		<input name="pageup" type="button" id="7" value="1" code="308" />
		<input name="r2" type="button" id="14" value="1" code="315" />
		<input name="r3" type="button" id="15" value="1" code="316" />
		<input name="right" type="hat" id="0" value="2" />
		<input name="select" type="button" id="9" value="1" code="310" />
		<input name="start" type="button" id="10" value="1" code="311" />
		<input name="up" type="hat" id="0" value="1" />
		<input name="x" type="button" id="6" value="1" code="307" />
		<input name="y" type="button" id="5" value="1" code="306" />`
	`</inputConfig>`

**Comment by Burnoh** [comment]
how are fixed retroachievements supposed to work? it still fails to boot a game when that feature is enabled in batocera settings.

**Reply by gkperkins** [reply]
I'm having the same problem, when enabled in ES I can go to the RetroAcheivements page in settings and see my achievements so I know the credentials are right, but games just crash when trying to start them. 

I've instead tried disabling in ES and enabling it in RA, but for some reason even though I can see it is enabled in the RA config games always start with it disabled. The only way I've got it to work is closing the game within RA, enabling Achievements, and then starting the game again within RA (not a great experience).

Regardless, the effort to get Batocera working is much appreciated, this seems to be moving along quite fast.

**Reply by gkperkins** [reply]
I was able to figure out the issue, it's because of a process when launching libretro (RA) where a python script is trying to check for an internet connection. It uses a bash command `timeout` to timeout the ping it does, but the problem is that timeout doesn't exist on the distro. My solution for now was to remove that timeout and now I'm not having problems (at least with RA emulators, haven't tried any standalone).

The file in question is located at `/usr/lib/python3.11/site-packages/configgen/generators/libretro/libretroConfig.py` line 62. 
I just removed `"timeout", "1", ` from the `cmd` list. I believe this would mean if you aren't connected to the internet there may be a delay in determining that, but it works for now, there is definitely a more elegant solution possible.

@acmeplus for your awareness

**Reply by acmeplus** [reply]
Thanks, that explains why it was working for me because I fixed the timeout yesterday and didn’t realize it was impacting that too.

**Reply by gkperkins** [reply]
Ah, yeah that makes perfect sense. In that case I'll just stick to my temp solution until next release. Thanks again for all your work!

**Reply by Burnoh** [reply]
I momentarily fixed it with the usual config change already reported and linked in the changelog above. 

Thanks for the overall work on this! 
This is my first version I'm trying as I received my rg35xx+ a couple days ago. All it's missing is the sleep function if you ask me 😄

**Comment by KingLJN** [comment]
Great work!

In my testing I found some issues:
1. Mupen64 RetroArch menu seems to have issues. The notification shows a bunch of weird colors which is sign that it might be using the wrong GPU driver (Similar to what happens on stock OS with the GPU driver switched to GLCORE from GL)
2. The issue I found on stock OS that's also on Batocera but not on MuOS: Changing some RetroArch settings (like CPU filters or Blargg NTSC in FCEUmm, maybe others) within the RetroArch menu might eat up RAM. If too much RAM gets taken up on Batocera, the whole system hangs, while on stock OS RetroArch shuts down and the system goes back to the main menu. 
![20240126003727](https://github.com/rg35xx-cfw/rg35xx-cfw.github.io/assets/47855056/99f2c055-af97-48ff-845d-15f805b9f96b)



Suggestion: 
Will later versions have media playback as an app that serves as a simple replacement for Kodi?

**Comment by pinkonmyshoe** [comment]
Love the work being put into this and using it as my daily driver since it pretty much ticks all the boxes. Just a couple questions:
- I'm struggling getting native pico-8 running (could be that I'm still a novice at this). Is this something that has been tested and works? Fake8 is fine for now but would love to get Splore up and running. 
- I'm unable to access the device from my computer. WiFi is working (retroachievements and downloads work fine + IP address is there) but internet status is saying 'not connected'. Could this be the issue?
